### PR TITLE
feat(renderer): add a formatting toolbar to the Source pane

### PR DESCRIPTION
**Pull request by jaden-cis**
## Summary
- The Source pane's CodeMirror editor had no way to format Markdown except typing raw syntax. Adds a blog-editor-style toolbar above it: a heading picker (Paragraph/H1–H6) plus bold, italic, strikethrough, horizontal rule, blockquote, bullet/numbered lists, checklist, link, inline code, and fenced code block.
- Toggle logic lives in pure functions (`markdownEdits.ts`) so it's unit-tested without mounting CodeMirror; `SourceEditor.tsx` exposes them as imperative commands (`toggleBold`, `toggleChecklist`, ...), and the new `SourceToolbar.tsx` wires buttons (icons from `Icons.tsx`) to those commands.
- Blank-line handling: a multi-line selection leaves blank lines inside it untouched (so intentional paragraph breaks survive a bulk toggle), but a single blank line under the cursor is still targetable — clicking Bullet/Checklist/Quote on an empty line starts a fresh one instead of doing nothing.
- Checklist/list/quote toolbar buttons only write the Markdown syntax (`- [ ] `, etc.); inplan's markdown-it preview has no GFM task-list rendering, so a checkbox still shows as literal text — that's a separate, unscoped follow-up.

### Also bundled: an Image toolbar button, with its own asset-saving infrastructure
This PR's title undersold it — it also adds a second feature with real security surface, called out here per review feedback:
- **`asset:save` IPC handler** (`packages/app/src/main/index.ts`): the main process writes a picked image's bytes into a `<docname>.assets/` folder next to the open doc (same pattern as "Create Doc" — never the sidecar dir, so the image is git-trackable and the link stays valid for anyone who clones the repo). The write path is derived from the *trusted* open-doc path, not anything renderer-controlled; the extension is regex-sanitized and the filename is a machine timestamp (no user-controlled name) — a malicious renderer call can't path-traverse via this handler.
- **CSP widening**: `img-src` now explicitly allows `file:` (previously implicit via `default-src 'self'`, which doesn't cover a doc living outside the app's own bundle directory). Required for a saved image to actually render in the preview.
- **`markdown.ts` image-src resolution**: a relative image `src` is resolved against the doc's real absolute path and rewritten to a `file://` URL. Absolute/URL/scheme sources are left untouched. Known accepted residual risk: resolution isn't clamped to the doc's directory subtree, so a doc containing ``-shaped Markdown will still attempt to load that file. Impact is low (an `` can't read pixel bytes back into the page — canvas taint — and the existing HTML-escaping blocks attaching `onerror`/`onload`, so it isn't a practical existence oracle either), and the renderer package has no filesystem access of its own to establish a repo-root boundary to clamp against. Tracked as a fast-follow.
- **Image markdown uses the angle-bracket destination form** ``, not bare `` — `relPath` is `<docname>.assets/…`, so a doc named e.g. "Product Plan.md" makes that a path with a space, which the bare (unbracketed) destination syntax can't parse as a link at all.

## Test plan
- [x] `markdownEdits.test.ts` — pure toggle logic for all edit functions, including the blank-line edge cases and the angle-bracket image destination
- [x] `SourceEditor.formatting.test.tsx` — imperative commands dispatched against a real mounted CodeMirror instance, including a lock-gate regression test (no toolbar command mutates the doc while `editable=false`)
- [x] `App.sourceToolbar.test.tsx` — toolbar wiring, heading-dropdown active-state, image-button disabled state, and MIME→extension mapping (including `image/svg+xml` → `svg`, not the invalid bare `svg+xml`)
- [x] `markdown.test.ts` — image-src `file://` resolution, including the angle-bracket/percent-encoding round-trip (asserts no double-encoding, e.g. no `%2520`)
- [x] Full renderer suite passing, typecheck clean across all workspaces
- [x] Verified against the real Electron app via Playwright `_electron`: toolbar renders, heading dropdown applies/toggles H1–H6, bold/checklist/list/quote work on both populated and empty lines, image button saves+renders correctly including for a doc name containing a space

<a href="https://cubic.dev/pr/melly-lgtm/inplan/pull/64?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added a Markdown formatting toolbar to the source editor.
  * Supports headings, text styles, quotes, lists, checklists, links, horizontal rules, inline code, and code blocks.
  * Added image insertion with local asset storage and relative image links.
  * Relative images now load correctly from their document location.
  * Added localized labels and responsive toolbar styling.

* **Tests**
  * Added coverage for formatting controls, heading behavior, image insertion, and relative Markdown images.

### CODE REVIEW

**Review by melly-lgtm — COMMENTED** [review]
Reviewed the full diff (17 files, ~1k lines). The formatting-toolbar core is well-built; my main asks are around the *second* feature this PR quietly bundles — on-disk image assets, a CSP widening, and `file://` image loading — which carries the real risk and isn't mentioned in the description.

### What's strong
- **Pure, DOM-free edit functions** (`markdownEdits.ts`) unit-tested independently of CodeMirror, then exposed as imperative commands mounted against a *real* CodeMirror instance. Right seam, real coverage including the blank-line edge cases.
- **`asset:save` is well-guarded against IPC-side path traversal**: `ext` is regex-sanitized, the filename is a machine timestamp (no user-controlled name), and the write dir derives from the *trusted* open-doc path — a malicious renderer call can't escape it.
- Image resolver correctly skips absolute/scheme/`//` srcs, handles the Windows drive-letter `file:///C:/…` form, and uses `encodeURI` (not `Component`). Web guard is clean (`disabled || !onPickImage`, so no crash where there's no filesystem).

### Findings

**1. [Correctness — should fix] Unescaped image paths break for common doc names.** `imageEdit` emits `` raw, and `relPath` is `<docname>.assets/…` derived from the user's `.md` filename. A doc named `Product Plan.md` or `My Plan (v2).md` yields a path with a space or paren, which markdown-it won't parse as an image URL — the link breaks at the moment of insertion. The new tests only cover space-free paths (`plan.assets/x.png`), so it's uncaught. Fix: emit the angle-bracket destination `` (markdown-it allows spaces + parens there); the read-side resolver still gets the clean path since markdown-it strips `<>` before `attrGet("src")`.

**2. [Security — low / defense-in-depth] Relative image resolution can escape the doc dir via `..`.** The new image rule in `markdown.ts` rewrites any bare relative src to `file://` + `resolveDocPath`, which normalizes `..` but doesn't clamp to the doc's subtree. Given inplan's threat model (docs co-authored with an agent and cloned from repos), a doc containing `` makes the app load an arbitrary local file. Impact is low — an `` can't read the bytes back (canvas taint), and #48's HTML-escaping prevents attaching `onerror`/`onload`, so it isn't even a practical existence oracle — but it's new surface. Worth clamping resolved image paths to the doc directory / repo root, or at least noting the decision. Reasonable as a fast-follow.

**3. [Process/docs — should fix] The description omits the image feature entirely.** It reads as a pure toolbar PR, but it also adds an `asset:save` IPC that writes files into the repo, a preload bridge, a CSP widening (`img-src file:`), and `file://` loading. The CSP change and main-process write are the most security-relevant parts of the diff — please spell them out in the description (or split them out) so reviewers and future readers see them.

**4. [Minor/robustness] `docPathRef.current` in a `useMemo` dependency array (App.tsx) is a React smell.** Refs are mutable and don't trigger re-render; the memo only recomputes because `doc.body` changes alongside docPath on navigation. Fragile if that coupling ever breaks — prefer threading docPath as state/prop.

**5. [Nits]** empty alt text `` (a11y); no size bound on `bytes` in `asset:save` (Typora-style "binaries into git" is a design choice worth a line of docs); the image button is correctly *disabled* on web but could just be hidden; the checklist button writes `- [ ]` that the preview renders as literal text (acknowledged follow-up — fine, mildly confusing UX).

### Net
Please address **#1 (path escaping)** and **#3 (description/CSP call-out)** before merge; **#2** is a fine fast-follow. Nice, clean core otherwise.

**Comment by jaden-cis** [comment]
Thanks for the thorough review — addressed in 4cefd16:

1. **Path escaping (fixed)** — `imageEdit` now emits `` (angle-bracket destination) instead of the bare form, so a doc name with a space/paren no longer breaks the link. Also had to fix `markdown.ts`'s resolver to `decodeURI` before re-encoding, since markdown-it already percent-encodes the destination at parse time — without that it double-encoded (`%20` → `%2520`). Verified against the real app with a "Product Plan.md" doc name.
2. **`..` path traversal (accepted as fast-follow, per your own read)** — agreed on the low practical impact, and the renderer package has no filesystem access of its own to establish a repo-root boundary to clamp against. Noted in the PR description for visibility.
3. **Description (updated)** — added a section spelling out the asset:save IPC, the CSP `img-src file:` widening, and the image-src `file://` resolution, so it's not buried in a PR titled "add a formatting toolbar."
4. **`docPathRef.current` in the `useMemo` deps (fixed)** — dropped it from the array with a comment explaining why `doc.body` alone already covers it.

Nits (#5) left as-is for now — happy to follow up on any of those if you'd rather they land in this PR.

**Comment by jaden-cis** [comment]
@coderabbitai review

File: packages/renderer/src/markdownEdits.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  const allPrefixed = targetable.length > 0 && targetable.every((l) => l.startsWith(prefix));
+  const insert = lines
+    .map((l) => {
+      if (skipBlank && l.length === 0) return l;
+      if (allPrefixed) return l.startsWith(prefix) ? l.slice(prefix.length) : l;
+      return l.startsWith(prefix) ? l : prefix + l;
+    })
+    .join("\n");
+  return { changes: { from: start, to: end, insert }, selection: { anchor: start, head: start + insert.length } };
+}
```
[the remark this answers is not in the document]
Comment (melly-lgtm) [reply packages/renderer/src/markdownEdits.ts]: This is already implemented on the current branch — the checklist call site passes exactly the proposed `prefixPattern`:

```ts
toggleChecklist() { withSelection((t, f, to) => linePrefixEdit(t, f, to, "- [ ] ", /^-\s\[[ xX]\]\s/)); }
```

Tracing the repro on current code: `"- [x] Task"` → the pattern matches `"- [x] "` (len 6) → `allPrefixed` true → `slice(6)` → `"Task"`. So it toggles off correctly, not `"- [ ] - [x] Task"`. The finding was against an earlier revision. Resolving.

**Review by melly-lgtm — APPROVED** [review]
Approving. All review findings addressed: image paths use the angle-bracket destination form (spaces/parens in doc names now work, with a round-trip test), the asset-saving IPC + `img-src file:` CSP widening are documented in the description, the `..`-resolution risk is captured as an accepted low-impact fast-follow, and the docPath `useMemo` smell is fixed. The one open bot thread (checklist toggle) was stale — the current code already implements its proposed `prefixPattern` fix. Clean toggle logic, good test coverage, CI green.
